packages/ERTP/src/issuer.js

```diff
@@ -0,0 +1,240 @@
+// Copyright (C) 2019 Agoric, under Apache License 2.0
+
+import harden from '@agoric/harden';
+import { assert, details } from '@agoric/assert';
+import makeStore from '@agoric/weak-store';
+import { E } from '@agoric/eventual-send';
+
+import makeAmountMath from './amountMath';
+
+function produceIssuer(allegedName, mathHelpersName = 'nat') {
+  assert(allegedName, details`allegedName must be truthy: ${allegedName}`);
+
+  const brand = harden({
+    // eslint-disable-next-line no-use-before-define
+    isMyIssuer: allegedIssuer => allegedIssuer === issuer,
+    allegedName: () => allegedName,
+  });
+
+  const amountMath = makeAmountMath(brand, mathHelpersName);
+
+  const paymentLedger = makeStore('payment');
+  const purseLedger = makeStore('purse');
+
+  const makePayment = () =>
+    harden({
+      allegedBrand: () => brand,
+    });
+
+  // Methods like deposit() have an optional second parameter `amount`
+  // which, if present, is supposed to be equal to the balance of the
+  // payment. This helper function does that check.
+  const assertAmountEqual = (paymentBalance, amount) => {
+    if (amount !== undefined) {
+      assert(
+        amountMath.isEqual(amount, paymentBalance),
+        details`payment balance ${paymentBalance} must equal amount ${amount}`,
+      );
+    }
+  };
+
+  const makePurse = () => {
+    const purse = harden({
+      deposit: (paymentP, optAmount = undefined) => {
+        const srcPayment = E.unwrap(paymentP);
+        const srcPaymentBalance = paymentLedger.get(srcPayment);
+        assertAmountEqual(srcPaymentBalance, optAmount);
```

Comment:
Just to check my understanding, you think that optAmount could be used dangerously for something other than its purpose of being used in assertAmountEqual?

Reply:
If its pre-coerced form were use twice under the assumption they agree, but if it made clever use of, for example, getters or proxies to successfully coerce to something valid and therefore stable one time, and successfully coerce to a different valid but stable thing another time, that might enable an attack, yes.

In this case, the attack does not look possible, because nothing is done with the uncoerced amount other than coercing it once. But it is a subtlety of the code easily lost under maintenance if unmentioned.